### PR TITLE
Implement better build action.

### PR DESCRIPTION
**Pull request by sbhadr**
**This PR does the following:**

Implements tests for Ubuntu, Windows, Macos platforms (@latest).
Implements tests for JDK support from 12 to 16.

Github Actions or rather the specific function of this PR is to better identify surface-level bad commits to prevent polluting production with instability, incompatibility or performance issues. This is not a substitute for complex error checking as a whole. Linter may or may not be included.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          uses: actions/setup-java@v2
+          with:
+            java-version: ${{ matrix.java }}
+        - name: Build Gradle
+          if: ${{ runner.os != 'Windows' }}
+          run: chmod +x ./gradew
+        - name: Build Repository
+          run: ./gradlew build
+        - name: Build Artifacts
+          if: ${{ runner.os == 'Linux' && matrix.java == '16' }}
```
Comment (crazyroo1) [inline .github/workflows/build.yml]: Should this build artifacts phase be in this action? Or should we split out the build testing from the artifact uploading?

Comment (sbhadr) [reply .github/workflows/build.yml]: I think perhaps later on we should split it out. For now, it's pretty simple to consolidate.

**Comment by sbhadr** [comment]
Currently blocked by this:
image

Don't approve. Attempting to fix.

File: .github/workflows/build.yml (hunk 1 above)
Comment (crazyroo1) [reply .github/workflows/build.yml]: Understood. LGTM

**Comment by crazyroo1** [comment]
De-indent the whole `steps` block, so it's in line with `strategy`

**Comment by sbhadr** [comment]
> De-indent the whole `steps` block, so it's in line with `strategy`

I figured this was it. Checking the status now.

**Comment by sbhadr** [comment]
Should work now. Edit: Builds show up. They fail on default. Identifying cause.

**Comment by sbhadr** [comment]
Distribution is required. Fixing.

**Comment by crazyroo1** [comment]
Ah, the joys of CI

**Comment by sbhadr** [comment]
Error on the Build Gradle phase. Looking into it. Seems to be a directory/chmod error code 1.

**Comment by sbhadr** [comment]
Going to eat dinner. I know the issue. I will fix ~30 minutes.

**Comment by crazyroo1** [comment]
Looks like java 16 is required for loom

**Comment by sbhadr** [comment]
Yup. Seems to be fine now. ``fabric-loom`` supports 16 or higher by default. We could technically get around this for backwards support for later versions. Not sure if it's worth it though; at least for now.

**Comment by crazyroo1** [comment]
Yeah, not sure of the implications of building with 16. A good question would be are older versions of java barred from running the mod, or is it simply a buildtime requirement with no runtime issues?

Of course, I don't expect us to know the answer without testing - and even then, how much do we care 😅

**Review by crazyroo1 — APPROVED** [review]

**Comment by sbhadr** [comment]
> Yeah, not sure of the implications of building with 16. A good question would be are older versions of java barred from running the mod, or is it simply a buildtime requirement with no runtime issues?
> 
> Of course, I don't expect us to know the answer without testing - and even then, how much do we care 😅

It would probably be wiser per-se to just assume that anything lower than Java 16 isn't supported. Once we have something ironed out, we can look into backwards support for the other Java versions. Though, I wonder which Minecraft versions Java 16 is supported with. Is it just 1.17.x or 1.16.x as well? 

Continuing the discussion. Merging.

**Comment by sbhadr** [comment]
Not sure how accurate this is but,
image

**Pull request merged**